### PR TITLE
feat: get statusbar height (top-inset)

### LINKED ISSUE #112: how to get statusbar height
could I get statusbar height with this plugin?

### PR DESCRIPTION
**Pull request by XxAdi101xX**
### Platforms affected
The main platform that this is desired for is Android. But this enhancement works on iOS as well.

### Motivation and Context

This adds a function that allows us to get the height of the status bar. This is inspired from this issue https://github.com/ionic-team/ionic/issues/17927 where the safe-inset paramters are not respected for Android devices so there is no way to get the statusbar height to enable appropriate styles to support app overlays.

This resolves #112 .

### Testing

With testing, the current android implementation AND iOS implementation along with the javascript bridge is working as intended! On an iPhone X, we get a value of 44px as intended. And on an android phone with and without a notch, we get the respective heights of statusbar. Required tests have been added.

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [x] I added automated test coverage as appropriate for this change
- [x] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] I've updated the documentation if necessary

### CODE REVIEW

File: src/browser/StatusBarProxy.js
Comment (XxAdi101xX) [inline src/browser/StatusBarProxy.js]: most of these changes are spacing issues, I've only added a default height of 0

**Comment by XxAdi101xX** [comment]
The code has been tested on both iOS and Android devices. Will add some tests next and update the docs as well. The tests seem to be CI tests seem to be failing because xCode isn't installed on the test pipelines and hence those need to be updated by the owner of the repo.

**Comment by XxAdi101xX** [comment]
The PR is ready for review.

**Comment by chavesgu** [comment]
What about your method for getting the height of the status bar，I think it's a necessary method for `cordova-plugin-statusbar`

**Comment by chavesgu** [comment]
How compatible is your method？ ios 9？  android 4 or 5 above？

**Comment by XxAdi101xX** [comment]
@chavesgu I have implemented the method for getting the height of the statusbar for both iOS and android. iOS compatibility does work with ios 9 since the method to get the statusbar on iOS was already used to implement some of the other methods in this repository already so it's just exposing a method that was always used. For Android, Android 4 above is supported. See https://stackoverflow.com/questions/29398929/how-get-height-of-the-status-bar-and-soft-key-buttons-bar. Overall, the solution is very compatible with older versions.

**Comment by XxAdi101xX** [comment]
@janpio hi, it's been a while since I've submitted this pr. Do you have any information regarding whether this is in a queue to be reviewed, or what the general status is in general. Thanks!

**Comment by breautek** [comment]
Things have changed since your PR and there is an outstanding merge conflict. Once that is resolved, I can look at the android side of things. Just tag me and I'll try to look at it within a week... and if I don't then feel free to ping me again.

**Comment by XxAdi101xX** [comment]
@breautek thanks for reaching out, I just resolved the merge conflict with the `README.md` file and there are no more conflicts. Please take a look whenever you can, thanks :)

**Comment by XxAdi101xX** [comment]
Hi @breautek, here's the friendly reminder that you requested! Please take a look at my changes whenever you can :)

File: src/ios/CDVStatusBar.m
Comment (breautek) [inline src/ios/CDVStatusBar.m]: Someone who has a bit more knowledgeable about iOS native development than me might be able shed more like but I think this `setKeepCallbackAsBool` is intended to be used to keep a reference of the callback so it can be invoked several times? Is this necessary here?

**Comment by breautek** [comment]
I think android looks fine. I'm not very knowledgeable on iOS development, but I think the `setKeepCallbackAsBool` usage might be out of place.

**Comment by pcsantana** [comment]
This is very useful, when could you release this functionality, please?

**Comment by EKashpersky** [comment]
@breautek I think we can learn if it doesn't work really fine on iOS faster, if the MR is accepted, because people are going to use this

**Comment by marpstar** [comment]
@breautek FWIW, we are running this branch in a production application on iOS and Android without issues.

**Comment by EKashpersky** [comment]
@breautek please notify if you ain't going to merge that.

**Comment by jenerse** [comment]
@breautek This is a useful functionality. Do you have plan to merge it? Thanks.

**Comment by breautek** [comment]
Before this can be merged we will need:
- https://github.com/apache/cordova-plugin-statusbar/pull/134#discussion_r350545105 to be addressed
- A second PMC member to vote in favour.

File: src/android/StatusBar.java
Comment (SchnWalter) [inline src/android/StatusBar.java:222]: Tested on a v27 emulator and on a v29 device, it works; Thanks!

I think that this feature will be used a lot to expose the height in a CSS variable, and I think that it might make sense to expose it directly in the plugin. I'm currently setting the CSS variable like this:
```js
const onSuccess = (height: number): void => {
  const roundedHeight = Math.ceil(height);
  if (isNaN(roundedHeight)) {
    document.documentElement.style.setProperty('--cordova-status-bar-height', '0px');
    console.warn('Failed to determine the StatusBar height.')
  } else {
    document.documentElement.style.setProperty('--cordova-status-bar-height', `${height}px`);
  }
};
```
And in CSS:
```css
.my-toolbar-container {
    padding-top: calc(8px + var(--cordova-status-bar-height));
}
```
But in order for this to work properly in all cases, the CSS variable value needs to be updated on each screen orientationchange. Because, in some cases, the StatusBar height depends on the height of the screen notch and if you rotate the screen, then the status bar height might be different depending if you have a notch or not.

It would be nice if the user could just set a:
```
<preference name="StatusBarHeightCSSVariable" value="true" />
```
In order to use a CSS variable directly, without having to write any JS in order to fetch the height and to subscribe to the screen orientation changes.

Comment (0x253) [reply src/android/StatusBar.java:222]: Perhaps https://developer.mozilla.org/en-US/docs/Web/API/Window/orientationchange_event could help to detect the screen orientation change and recompute the statusbar height?

**Comment by 0x253** [comment]
> @breautek FWIW, we are running this branch in a production application on iOS and Android without issues.

Tested this changes on various devices (iOS and Android), always computes the correct statusbar height. Will use it in our production apps.

File: src/android/StatusBar.java
Comment (SchnWalter) [reply src/android/StatusBar.java:222]: That works, but if the CSS variable is defined, you don't have to care about anything. You just use it as it is. 

This is the same approach that the `iOS` guys took in order to avoid issues with screen notches. The've added those `env(safe-area-inset-*)` custom CSS "environment variables".

* https://developer.mozilla.org/en-US/docs/Web/CSS/env

And they get updated when you rotate the screen and the screen notch is on a different screen edge.

Comment (0x253) [reply src/android/StatusBar.java:222]: I know, I use the Safari `safe-area-inset-*` variables since iPhone X. What you are suggesting is an Android Chrome bugfix plugin. Basically, a Corodva plugin for Android without an exposed JavaScript interface, that sets the `safe-area-inset-*` env variables in the Android WebView's document. The plugin could also handle orientation changes. I think the Statusbar plugin is not the correct place to fix the missing env variables of Chrome.

Comment (SchnWalter) [reply src/android/StatusBar.java:222]: It's not a Chrome Bugfix, because that standard doesn't exist yet, it's still a draft, and from what I can tell, you can't provide those `env()` variables in a polyfill.

And I'm also not talking about implementing the `env(safe-area-inset-*)` as part of the StatusBar, I'm just saying that the status bar should always expose it's own height.

And now that I think about it, the CSS variable should have a fixed value and it should not depend on orientation changes.

**Comment by vadimwe** [comment]
Cant install ionic cordova plugin add...

**Comment by SchnWalter** [comment]
@stevengill, @erisu, is there something we can help with to get some of these PRs merged? Would you like us to provide some code cleanup? Provide rebased branches (with proper attribution to the original author) or with more granular changes and some extra tests?

**Comment by jglesner** [comment]
This would be a really helpful feature to programmatically adjust layouts after orientation changes.  Feels like this should be a feature of the statusbar plugin. Its a bit disappointing that this was contributed well over a year ago, but never merged.  Any chance this could be closed out?

File: src/ios/CDVStatusBar.m
Comment (dertieran) [reply src/ios/CDVStatusBar.m:202]: Yes that is the intend, otherwise the callback will be deleted on the js side.
https://github.com/apache/cordova-js/blob/0608bf932289c143405e90dc8edd3449f78f5fdc/src/cordova.js#L230-L233

Comment (SchnWalter) [reply src/ios/CDVStatusBar.m:202]: Thanks for the feedback.

@breautek, if you want to merge this, I've created a rebased version of this branch, you can merge that one if you want, (with attribution to the original authors)

> https://github.com/SchnWalter/cordova-plugin-statusbar/tree/134-rebased/expose-StatusBar-height

**Review by breautek — APPROVED** [review]
lgtm :+1:

Thank you @SchnWalter for taking the time to rebase this PR.

Tested on android emulator with:

- Solid Statusbar
- Overlayed Statusbar
- With display cutout 

Each appeared to return the proper size in pixels.

I don't have a mac to test iOS, so it would be great if one of the other PMC members is able to give it a quick test on their macs.

File: src/ios/CDVStatusBar.m
Comment (dertieran) [inline src/ios/CDVStatusBar.m:206]: Just realized my [comment](https://github.com/apache/cordova-plugin-statusbar/pull/134#discussion_r514301773) might be misleading,`setKeepCallbackAsBool` should **not** be set to `YES`,
because this could causes a memory leak because the callback handler is never cleaned up.
```suggestion
```

**Review by dertieran — CHANGES_REQUESTED** [review]

**Comment by jenerse** [comment]
@breautek When will this branch be merged?

**Comment by skks1212** [comment]
Really need this feature for my app. When will it be merged?

**Comment by mianfg** [comment]
Hello, I have seen that you have implemented this in production. I am using Ionic, and I cannot find a way to use this function (as Ionic's StatusBar library does not have this funcion implemented). Let me describe the issue.

If I try to use StatusBar via
```ts
declare let window any;
```
it seems that the height function is not defined, as printing `window.StatusBar` gives me:
```
StatusBar:  
backgroundColorByHexString: ƒ (hexString)
backgroundColorByName: ƒ (colorname)
hide: ƒ ()
isVisible: true
overlaysWebView: ƒ (doOverlay)
show: ƒ ()
styleBlackOpaque: ƒ ()
styleBlackTranslucent: ƒ ()
styleDefault: ƒ ()
styleLightContent: ƒ ()
__proto__: Object
```
Just to clarify, I am using `cordova` via `ionic cordova run android` and using the device inspector to view the console output.

Sorry if this is a somewhat naive question, but I have been reading about this for days and I cannot seem to find any way to make this function callable. Any help would be appreciated!!!

**Comment by nielsjakob** [comment]
> Really need this feature for my app. When will it be merged? @breautek

Me too :)

**Comment by ollm** [comment]
Tested on android emulator and seems to work perfectly, except in Android's Multi-Window mode, for instance, when in portrait orientation, the status bar might be visible if I'm the top app, but won't be visible when I'm the bottom app.

Maybe it can be implemented like this in Javascript:

``` javascript
StatusBar.height(onSuccess, onError);

const onSuccess = (height, statusBarHidden) => {
    // do something with the statusbar height here
}

const onError = (error) => {
    // handle error state; should usually not ever activate
}
```

**Comment by piwpio** [comment]
> Really need this feature for my app. When will it be merged? @breautek

Same here, can't wait to be merged :)

**Comment by XxAdi101xX** [comment]
I'm surprised that this has not been approved yet after all this time. I had started to use this custom branch for my projects but it would be ideal if it's part of the main branch given the demand.

**Comment by coloz** [comment]
Why hasn't this pull been merged for so long?

**Comment by breautek** [comment]
> Is this PR going to be merged? I see there are 3 pending reviews, I would like to use this with my ionic cordova app. So it would be nice if this could be merged into master so that ionic native can update to the latest package.
> 
> Please let us know what will happen to this...

PR has gone stale with a merge conflict and no activity from other PMC members. So I'm going to just close this PR. Sorry @XxAdi101xX

I do have a workaround however that probably fits your needs better than what this PR attempted. The company I work for implemented a [insets](https://github.com/totalpaveinc/cordova-plugin-insets) plugin. Like this PR, `@totalpave/cordova-plugin-insets` is an android-only plugin so if you support iOS then you'll need to platform guard around the usages of the plugin. Safe area insets via CSS works on iOS and that will be the preferred approach. This insets plugin just provides a fallback for android.

The plugin is agnostic to how you want to apply the insets, it just gives you the information, it's up to you how you want to glue it to your application. For example, you could create [custom CSS variables](https://github.com/apache/cordova-plugin-statusbar/pull/134#discussion_r413468163) with the returned values so that the values can be used within your CSS. (However I found that setting CSS variables is slow, so I update my viewport div directly in JS instead where necessary).

The plugin returns the inset information to count for both system bars as well as display cutouts / islands and it will update you via a callback should they change, for example for orienation changes or if the user changes settings that affects the system bars.

Disclaimer: I am part-author of the insets plugin.

**Pull request closed without merging**
